### DISCUSSION TITLE
What is the mean magnitude of an experiment?

### DISCUSSION THREAD

**Discussion by rodolfocarobene** [qiboteam/qibocal#1310] [General]
Hi everyone,
the problem is a bit inter-library (qibocal-qibolab-qibosoq), but I am opening here since it is of characterization nature.

I was running a resonator spectroscopy followed by a qubit spectroscopy and I noticed that the "baseline" of the qubit spectroscopy was different than the corresponding value of the resonator spectroscopy. Tracing back the problem, I think the difference was that in qibocal the mean magnitude across multiple shots is computed as the mean of all the magnitudes (this was used in the res spec), while in qibosoq I was following the "QICK practice" of computing the mean value of I, the mean value of Q and only than compute the magnitude.

Now, I am not 100% sure of what is the best/correct way but for sure we need to stick to one of the two options otherwise we get incoherent values. At the moment, it seems to me that is more reasonable to consider the mean of I and the mean of Q and than compute phase/magnitude.

What are your thoughts on this? Maybe you already discussed it and already reached an agreement, in that case I'll change my side of the libraries

**Comment by alecandido** [comment]
From my side, there is not any existing agreement about this topic.

Also, in general I was thinking to phase out the magnitude and phase when not strictly needed: in resonator spectroscopy, we could fit separately the two components of the resonance. But the `s21` fit by @marcogobbo is also doing the fitting both at the same time - and I'm not even that sure we would need anything else, once we decide that the `s21` fit is sufficiently reliable.

For Rabi and the other qubit-signal routines, I was planning to switch to PCA #1292. However, for spectroscopies we may still need, since they do not have any *state* interpretation. Even if we have the circle picture for the resonator (all in all, I've just never thought about how a qubit spectroscopy should look like in the IQ plane - and it could be simpler to interpret also other sweeps just in magnitude, as the flux dependencies)

**Reply by rodolfocarobene** [reply]
Thanks for the comment! Let's hear if someone else has preferences on the mean method.

Regarding PCA I am not sure it's a good idea because you get a bit away from a physical quantity and it could be more complex to interpret and debug, if something weird appears.

Regarding your first point (removing magnitude and phase when not strictly needed), what do you mean? Just remove one of the two or what? Sometime, even if the fit is not considering it, the plot of the phase has valuable information

**Reply by alecandido** [reply]
> Regarding PCA I am not sure it's a good idea because you get a bit away from a physical quantity and it could be more complex to interpret and debug, if something weird appears.

For Rabi oscillations, the physical quantities are not actually meaningful. If your centroids are in different quadrants, the oscillations in magnitude are not even monotonic, leading to very misleading plots.
The alternative we considered before was to just plot the quadratures, but then the problem is what to fit.
But the one value which you are interested in fitting is certainly the primary axis of oscillation (connecting the two centroids). And if there is no clear primary axis, you're really failing in seeing Rabi oscillations.
This is the same for all experiments relying on the qubit state (e.g. including T1 or T2 experiments).

In any case, on top of the PCA-decomposed dimensions, I want to also add the visualizations of points in a two-dimensional plane. Which, at that point, it will be the IQ plane itself.

> Regarding your first point (removing magnitude and phase when not strictly needed), what do you mean? Just remove one of the two or what? Sometime, even if the fit is not considering it, the plot of the phase has valuable information

As said: fitting the coordinate over the primary axis (as opposed to magnitude) and plot both coordinates, and the 2D scatter plot. That should convey all the required information.

**Comment by alecandido** [comment]
@rodolfocarobene since we do not have any clear action plan, I moved this into a discussion. As soon as we reach an agreement, we can open an issue with the result of this process.

**Comment by rodolfocarobene** [comment]
After some thought, I would say that the two approaches are not equivalent. Taking the mean of the magnitudes mixes in a noise-dependent positive bias, so the baseline depends on the shot noise level. Averaging I and Q first and then computing magnitude gives a cleaner estimate of the coherent response and keeps baselines consistent across experiments and libraries. This also explains the discrepancy I was seeing and the solution is the standard treatment in microwave heterodyne detection.

For the phase, the situation is similar: the meaningful quantity is the phase of the averaged complex signal, i.e. arg(⟨I⟩ + i⟨Q⟩). Averaging phases shot by shot is not equivalent and can behave badly in the presence of noise or wrapping.

Unless someone objects or adds further considerations, I will start a PR soon to standardize on averaging I and Q first, and only then computing magnitude and phase, and align the behavior across libraries.

**Reply by alecandido** [reply]
I did not disagree about the averaging order. But, what I mean is that the relevant noise is a 2D effect.
I.e. the most of the information you get into the 2D plane. And I/Q are just a reference frame in that plane - I see no reason to claim it to be the best.

Indeed, the "best" frame we are calibrating ourselves in the single-shot readout experiment. That's exactly where we calibrate the rotation and threshold, which is nothing else than another reference frame.

So, my point was that the PCA is the best approximation you could get to classification, without performing any SSRO experiment first. Even if you're running a Rabi (or, for what is worth, a qubit spectroscopy) you will see the signal moving between the positions related to the two states of the qubits. And the PCA will be the closest approximant frame to the best linear classifier. And it's just a linear (actually affine) combination of I and Q. In the worst case, it will be as random as the I/Q one.

The bad part of the magnitude and phase coordinates are just that they are not linear in I and Q, and averaging in that space may cause some biases, as you explained. But I do not see how you could introduce a similar bias with an affine combination (which will anyhow commute with the average, since they are both *linear*).

**Reply by rodolfocarobene** [reply]
Yes sorry, I am not really thinking about PCA at the moment because I thought this was a "second-step" to be cosnidered at a different time scale. It seems to me that the PCA implementation will require some more time, won't it?
In the meantime however, there is (potentially) some contrasting behaviour between libraries (qibocal and the instrument APIs) that would need a quick fix to be unified

**Reply by alecandido** [reply]
The main advantage of the quadratures averages is that you could do that on board.
But if you download the individual shots, it should be pretty simple to apply the [`PCA`](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA).

```py
# test data
import numpy as np
X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
# actual PCA application
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
pca.fit(X)
# print the results
print(pca.explained_variance_ratio_)
print(pca.singular_values_)
```

So, it should be pretty straightforward.

But, it's true that at this point I would add some additional plotting: the 2D plot itself, and possibly with the PCA axes drawn (while just replacing the magnitude and phase plot for the principal components).

In any case, it is fine to consider a step beyond. I had just the impression that you did not agree on that.

The only thing to assess is in which (signal) protocols showing which information. Sometimes the magnitude picture is actual more familiar than anything else (e.g. in the punchout or the dispersive shift experiments).
